### ISSUE TITLE
Any way to close a channel explicitly?

### ISSUE THREAD

**Issue by BusyJay** [crossbeam-rs/crossbeam#236]
crossbeam-channel can detect close when all sender or all receiver are closed. Is it possible to close a channel explicitly? For example, a channel may need to be closed eagerly when one of its other endpionts detects errors.

**Comment by stjepang** [comment]
> Is it possible to close a channel explicitly?

Not currently possible, but it'd be fairly easy to implement this feature.

> For example, a channel may need to be closed eagerly when one of its other endpionts detects errors.

Isn't the `Receiver` or `Sender` going to be dropped anyway when an error is detected? I'm just curious to understand better why channels must be explicitly disconnected in your use case.

**Comment by BusyJay** [comment]
> Isn't the Receiver or Sender going to be dropped anyway

In my case, there will be some cache that still holds a reference to the sender, and receiver closes only when all senders are closed. So the channel won't be closed immediately.

**Comment by stjepang** [comment]
I'm feeling slightly hesitant to add this feature because it makes the disconnection mechanism a little bit more complicated. But, if `futures-channel` [has this feature](https://docs.rs/futures-channel/0.2.1/futures_channel/mpsc/struct.Receiver.html#method.close), perhaps it's a no-brainer to add it to `crossbeam-channel`, too...

**Comment by gterzian** [comment]
Could this problem perhaps be solved with a side "shutdown" channel? In such a setup, on the receiving side, you select over your "main" channel, on which "normal" communication comes in, and when a message comes in on the other "shutdown" channel, you shutdown, which would include dropping the receiver of the "main" channel. 

I'd say you could even just have a `Msg::Shutdown` handled on the "main" channel directly, but I'm guessing you don't want the other senders to keep sending messages when the `Msg::Shutdown` has already been sent, but not yet received? @BusyJay 

A bigger challenge would be if you have multiple receivers, since the one message you would send on the "shutdown" channel would not reach all receivers? What to do in that case? Keep sending messages on the "shutdown" channel, until it is disconnected? Or, hide your multiple receivers/workers behind a single intermediary "router", who could then stop forwarding messages once something comes in on the "shutdown" channel...

Also note that the use case for `close` in `futures::channel` seems to be the mirror of this problem:

> If the Receiver is simply dropped, then it is possible for there to be messages still in the channel that will not be processed. As such, it is usually desirable to perform a "clean" shutdown. To do this, the receiver will first call close, which will prevent any further messages to be sent into the channel. Then, the receiver consumes the channel to completion, at which point the receiver can be dropped
https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.11/futures/channel/mpsc/index.html#clean-shutdown

I guess that use-case could also be dealt with using some sort of "reverse shutdown" channel, where sending a "normal" message on the "main" channel would consist of selecting over the "send" and a "recv" on a shutdown channel. Then the "main" receiver could, when shutting down, first send a message on the "reverse-shutdown" channel, then drain the "main" channel, and finally drop it?

**Comment by stjepang** [comment]
@gterzian Setting up an auxiliary channel for shutdown is a lot of work and it might be easy to forget to select over the main channel and the shutdown channel.

The use case for `Receiver::close()` in `futures::channel` you've mentioned is very compelling. I really like this close-and-drain pattern.

**Comment by gterzian** [comment]
I agree that an auxiliary channel is more complicated than calling a `stop` method, and I also think one could avoid adding features that could be implemented with basic building blocks, especially when such features could turn a channel into something that it is not(for example, introducing some concept of shared-state, or sync communication for an unbounded channel). 

My experience of using channels comes down to an "actor-like" form of concurrent programming, with async message passing, no shared-state, and having the various components in the system run logically "in parallel"(even though they will not all actually run fully in parallel due to physical restrictions of the machine). 

From my perspective, the idea of "eagerly closing" a channel, or ensuring errors are "immediately" propagated, points more towards a need for redesigning the system, as opposed to adding such features to the means of communication used by the system. Such redesign could focus on removing the need for errors to "propagate immediately" instead embracing the async nature of message passing. 

Although the `close` use-case mentioned in the futures doc is compelling, it also makes a lot of assumptions about the needs of a given system. If the component holding the receiver has decided to shut itself down, why would it care about messages not yet handled? A "clean" shutdown could as well be argued to be empathically not handling any further messages, even when those have already been "sent" by other components.

Other people are undoubtly using channels differently, especially in the context of futures which come with a very different executing context. That could be an argument against standardizing channel APIs, since the context of their use will be different.

**Comment by BurntSushi** [comment]
@stjepang It is typically much harder to argue against seemingly harmless API additions. I would instead kindly suggest that we flip the burden here. I think in this case, especially if `disconnect` is something that could be added in a backwards compatible manner, it would be nice to get more use cases that truly demand it. It introduces a secondary way of shutting down a channel distinct from causing all senders (or all receivers) to be dropped. It's hard to reason about exactly how a secondary shutdown mechanism might be negative, but I could certainly see it resulting in manifest complexity in larger systems. For example, I _believe_ that with today's API, if you know that there are still at least one live sender and one live receiver, then the channel is still connected. If I'm understanding the proposed feature, if you introduced a `disconnect` method, then this guarantee is no longer possible to make.

There are also potential footguns associated with introducing accessor methods like `is_disconnected`, where a decision based on it becomes stale and you wind up introducing a race. This is why, for example, `chan` never had `len` or `is_empty` methods.

**Comment by stjepang** [comment]
@BurntSushi That is fair, no need to rush with a decision here!

> For example, I _believe_ that with today's API, if you know that there are still at least one live sender and one live receiver, then the channel is still connected.

Yes - this is certainly an invariant one might rely on.

> There are also potential footguns associated with introducing accessor methods like `is_disconnected`, where a decision based on it becomes stale and you wind up introducing a race. This is why, for example, `chan` never had `len` or `is_empty` methods.

Well, `len` and `is_empty` are useful - for example, Servo uses these methods to compute the estimated memory usage of channels. Even Go has similar functions.

We could also argue in favor of `disconnect()` when it comes to races. Currently, there is no "clean" (non-racy) way to drain and disconnect a channel. Even if we do `for msg in receiver {}`, we can't guarantee there won't be messages left in the channel. That is because `send()`, `try_recv()`, and `drop(receiver)` are racy operations. I'm not even sure there's a way to "fix" `Receiver::into_iter()` to make it non-racy. But with `receiver.disconnect(); for msg in receiver {}` we could be sure the channel really gets drained.

**Comment by BurntSushi** [comment]
> Well, len and is_empty are useful - for example, Servo uses these methods to compute the estimated memory usage of channels. Even Go has similar functions.

Yes, I know. :-) I don't think I've ever used them myself, and if I ever saw them used (in Rust or Go code), they would be immediately flagged for closer inspection. I wrote `chan` without them to see if I could get away with it. But yeah, if it had gotten more popular, I'm certain I would have added them!

> Even if we do `for msg in receiver {}`, we can't guarantee there won't be messages left in the channel. That is because `send()`, `try_recv()`, and `drop(receiver)` are racy operations.

I think I might be missing something here. I had thought that the only way for `for msg in receiver {}` to terminate would only be if all messages had been received *and* if all senders were dropped. That is:

```
let (sender, receiver) = bounded(1);
sender.send(1).unwrap(); // non-blocking
drop(sender); // all senders are gone
assert!(1, receiver.iter().count()); // iterator still consumes everything
```

**Comment by stjepang** [comment]
Ugh, sorry, I got myself confused here! That was a bad example and what you said is correct. :)

So my point was supposed to be about termination from the receiver side as follows:

```rust
// Receiver is shutting down. Consume all remaining messages.
for msg in receiver.try_iter() {
    do_something(msg);
}

// Disconnect.
drop(receiver);
```

This is racy because messages could be pushed just before we disconnect.

However, with explicit `disconnect()` we can tell senders to stop producing work *before* we drain the channel:

```rust
// Disconnect.
receiver.disconnect();

// Receiver is shutting down. Consume all remaining messages.
for msg in receiver.try_iter() {
    do_something(msg);
}
```

Finally, one more thought - there are cases where `disconnect()` is more convenient than wrapping a `Receiver` into `Option` and `Cell` so that it can be dropped manually. Consider this:

```rust
struct Foo {
    receiver: Cell<Option<Receiver<Message>>>
}

impl Foo {
    fn try_receive_one(&self) -> Option<Message> {
        // Uglier than `self.receiver.try_recv().ok()`.
        let r = self.receiver.take();
        let msg = r.and_then(|r| r.try_recv().ok());
        self.receiver.set(r);
        msg
    }

    fn shutdown(&self) {
        // `self.receiver.disconnect()` would be very handy here.
        self.receiver.set(None);
    }
}
```

**Comment by BurntSushi** [comment]
@stjepang Ah ok! So I think I'm still missing something though, because even in your `try_iter()` example, my understanding is that the _only_ way for `drop(receiver)` to be executed is if all receivers have been dropped _and_ all remaining messages have been processed.

w.r.t. to your `Cell` example, I guess I don't feel particularly strong there. I wouldn't bat an eye at `Cell<Option<Receiver<Message>>>` other than to perhaps ask why not use `&mut self` without the `Cell`, and then use `self.receiver.take()` to drop it. (Or even the more general question of why it needs to be explicitly dropped in the first place, but I'm pre-supposing we've invented a use case for it, which kind of makes me a little uneasy because it's hard to reason about how common it is. It definitely _feels_ uncommon to me.)

**Comment by stjepang** [comment]
> my understanding is that the _only_ way for `drop(receiver)` to be executed is if all receivers have been dropped _and_ all remaining messages have been processed.

Hm, not sure I understand. Suppose there is only one receiver associated with the channel.

It is possible to execute `drop(receiver)` without consuming a message that gets sent from another thread *just* before we drop the receiver. That message will be stuck in the channel until it gets destroyed, which is a bit unfortunate.

**Comment by BurntSushi** [comment]
Errm, sorry, I meant, the only way for `drop(receiver)` to be executed is if all *senders* have been dropped and all remaining messages have been processed. Basically what I'm saying is that the termination semantics of the iterator itself should be preventing races.

**Comment by BurntSushi** [comment]
Oh, oof, I see. `TryIter` has different termination semantics in that it will quit as soon as it has to block for a message. To me, that's part of the deal here. There's no synchronization point to make sure you've received all messages. I understand now what you're saying about `self.disconnect()` enabling one to use the API this way, but I don't think I understand why it's desirable.

**Comment by stjepang** [comment]
@BurntSushi

> I understand now what you're saying about `self.disconnect()` enabling one to use the API this way, but I don't think I understand why it's desirable.

A possible use case is when the worker thread (or actor) decides to perform a clean shutdown routine. For example, when the Tokio runtime triggers shutdown, we want to:

1. Prevent new tasks from being sent.
2. Complete all remaining tasks.
3. Exit the worker thread.

The worker thread could do that by disconnecting the channel and then processing the remaining tasks. The alternative is to use a separate channel to signal that the worker is shutting down.

-----

I wonder if we could instead add a new method `drain()` (maybe named differently) that drops the receiver *and then* receives remaining messages.

```rust
impl<T> Receiver<T> {
    fn drain(self) -> Drain<T>;
}
```

This is the opposite of `into_iter()` because `IntoIter<T>` holds a real `Receiver<T>`, while `Drain<T>` would not.

@BusyJay What do you think, would adding the `drain()` method be a good compromise here?

**Comment by BusyJay** [comment]
> ...adding the drain() method be a good compromise here

What if there are multiple `Receiver`s? Will drain close those receivers too? There is only one receiver in my case, so no mater what the answer is, `drain` works for me.

**Comment by stjepang** [comment]
> Will drain close those receivers too?

No, it would just drop the receiver that has been turned into a drain iterator.

**Comment by wmanley** [comment]
In the discussion of `panic!` on `send` (context [here][1] and [here][2]) there was a lot of talk about requiring code to be correct by construction.  In particular there's the idea that `send` panics by default because code that could send to a disconnected channel is already deadlock-prone by construction (discussion on #314).

I've been doing some thinking about what "correct by construction" means, and if there's a straightforward set of patterns that can be applied that are correct.

@emk gave an example of structuring UNIX pipeline style code in this [reddit comment][3] and in [this github comment][4]: 

>     data_producer | data_transformer | data_sink

I've attempted to write robust example code implementing this, but I've had great difficulty creating something that isn't excessively verbose, but is also robust against errors and race free. I think verbosity is important because the verbose option isn't going to be the default option users reach for.  I think this applies both to crossbeam-channels and tokio channels.

One difficulty is that if `data_sink` gets EOF from its `Receiver` it doesn't know if it's because `data_transformer` finished sending data, or if it's because it encountered an unexpected error and the `Sender` was just dropped. I would propose adding a method:

    fn close(self);

and extending `RecvError` to be able to distinguish between disconnection because the send side called `Sender::close` and closing because the `Sender` was dropped.

Then the pattern is: do your work, and once you know that all the work is done successfully call `close`.  This is inspired by the C++ ["commit or rollback" idiom](http://www.gotw.ca/gotw/059.htm) where you call a function after doing fallible work to say that no failures occurred.

This is different than the explicit `disconnect` proposed above because it consumes the `Sender` rather than leaving a `Sender` around that can't be used.  AIUI the motivation for `mut&` is so you can avoid storing your `Sender`s as `Option<Sender>`, whereas the motivation for `close` is robustness.  Possibly I should raise this as a separate issue?

Problems with this approach:

* The semantics are clear for SPMC, but not so for MPMC.  Options:
    1. Calling close on one `Sender` should probably cause `send` on other `Sender`s of the same channel to `panic`.
    2. A new SPMC channel type could be made available with a `close` method, while MPMC continues to lack this ability
    3. All `Sender`s must call `close` for the receiver to be told that the channel was shutdown cleanly.
* It doesn't help to detect errors where the `Sender` is not dropped, like deadlocks/livelocks - but if a deadlock detection/recovery system is in place this approach could compliment it.

Alternatives and extensions: 

* Don't modify channel and send an enum with an explicit end value.  So instead of `Sender<T>` use `enum { Value(T), SuccessfulShutdown }`.  This has the disadvantage (or possible advantage?) that you can carry on calling `send(Value(xyz))` after calling `send(SuccessfulShutdown)`.
* Have a second error channel and use something like [`scopeguard`](https://crates.io/crates/scopeguard) to notify down that channel when something goes wrong, rather than calling close once everything's gone right.  I think a second channel possibly makes things much more verbose.  I also think calling a method on success is easier to understand than trying to distinguish between success and failure using scopes.
* Allow sending a payload with `close` so `Sender<T>` becomes `Sender<T, FinalT>` and `close(self)` becomes `close(self, value: FinalT)`. The return value from `recv` would also have to be modified. This way you can provide even more context to the receiver.  This would make dealing with MPMC even trickier, but could be worth the complexity.  This would be analogous to a [`Generator`][5] which has seperate `yield` and `return` types.

[1]: https://www.reddit.com/r/rust/comments/awh751/proposal_new_channels_for_rusts_standard_library/ 
[2]: https://www.reddit.com/r/rust/comments/azaw4m/in_nightly_rust_await_may_never_return_dropping/
[3]: https://www.reddit.com/r/rust/comments/awh751/proposal_new_channels_for_rusts_standard_library/eho5a3o/
[4]: https://github.com/crossbeam-rs/crossbeam/issues/314#issuecomment-469039662
[5]:

**Comment by emk** [comment]
> One difficulty is that if data_sink gets EOF from its Receiver it doesn't know if it's because data_transformer finished sending data, or if it's because it encountered an unexpected error and the Sender was just dropped.

One trick which helps a bit with this is declaring the channel to hold a `Result<T, Error>`, and when the sender encounters an error, forward it to the channel before closing.

This is not necessarily robust in the face of panics, however. On the other hand, I generally run with `panic = "abort"`, because I've seen plenty of Rust libraries fail in annoying ways when a background worker panics. But that's not a general solution.

**Comment by wmanley** [comment]
> One trick which helps a bit with this is declaring the channel to hold a `Result<T, Error>`, and when the sender encounters an error, forward it to the channel before closing.

Indeed.  This is a pattern I tried, but I've found it quite awkward to implement in practice.  It's tricky to ensure that you're catching all the appropriate errors from all the call sites to forward them, and as you say it doesn't help with panics.  Panics are where `scopeguard` can come in handy, but it's a lot less obvious what's happening then.  Using `scopeguard` to notify errors when `async` `Futures` are dropped would probably be possible too, but it's more work, so I would guess in the common case it probably wouldn't be implemented at all, let alone correctly.  As you know, you discuss this case in [your blog post "await may never return"](http://www.randomhacks.net/2019/03/09/in-nightly-rust-await-may-never-return/).

**Comment by emk** [comment]
Yeah, in my personal experience the sending code has often (fortunately!) been less than 20 lines of code, so it's possible to keep track of errors and panics. Let's assume that you're implementing something like my "pipes" example. In the more complicated cases—for example, if the sending end can be shut down by a timeout or a dropped future—then probably you'd want a type like:

```rust
enum Piped<T> {
    Message(T),
    Done,
}
```

...and an stream item type of `Result<Piped<T>, Error>`. If the channel closes without receiving `Ok(Message::Done)`, then it has failed. 

This has the advantage that you can wrap an existing channel sender & receiver, adding the behavior needed for this use case.

It's entirely possible that I'm overlooking something. I usually am. I personally find that channels are a surprisingly difficult concurrency primitive, at least once I get into these details.

**Comment by iesahin** [comment]
As another data point, I've decided to use `Option<T>` as message type to signal to `recv` that the channel is done, and there won't be more messages. Instead of using `try_recv` and sleeping the thread when we get `None`, this makes it possible to block the receiver thread/loop indefinitely, and quit gracefully when the main thread sends a `None` message.

**Comment by ryoqun** [comment]
heh, this is a fairly good and old issue. :) fyi, I'm creating a pr for this: #959

**Comment by justpretending** [comment]
As one who has come from Go, I miss this feature. It's customary in Go for the writer (sender) to close the channel. For example, a jobs channel will be closed when there are no more jobs. The receiver goroutines will consume everything, then detect that the channel has been closed, and exit. If we didn't close the channel, they wouldn't exit, thus creating a leak.

**Comment by justpretending** [comment]
@iesahin `Option<T>` isn't really a solution when there are multiple receivers.

**Comment by justpretending** [comment]
Ah, I see. One can just `drop()` the sender.
